File: leet_3005_max_freq.go
func maxFrequencyElements(nums []int) int {
    Solutions := []func([]int)int{
        // Space_heavy,
        Best_solution,
    }
    return Solutions[ 0 ](nums)
}

func Best_solution(nums []int) int {
    D := map[int]int{}
    top, times := -1, 0
    for _, n := range nums {
        D[n]++
        if D[n] > top {
            top = D[n]
            times = 0
        }
        if D[n] == top {
            times ++
        }
    }
    return top * times
}

func Space_heavy(nums []int) int {

    D := map[int]int{}
    for _, n := range nums {
        D[n]++
    }
    N := len(D)
    A := make([]int, N)
    i := 0
    for _, v := range D {
        A[i] = v
        i++
    }
    sort.SliceStable(A, func (l, r int) bool {
        return l < r
    })
    top, res := A[0], 0
    for _, v := range A {
        if v == top {
            res += v
        } else {
            break
        }
    }
    return res
}

File: leet_3005_max_freq.cc
class Solution {
public:
    using PII = pair<int, int>;

    int maxFrequencyElements(vector<int>& nums) {
        vector<int(*)(vector<int>&)> Solutions {
            Best_complexity,
            // Space_heavy,
        };
        return Solutions[0](nums);
    }

    static int Best_complexity(vector<int>& nums) {
        unordered_map<int, int> D;
        int times = 0, top = -1;
        for (int & n: nums)
        {
            D[n]++;
            if (D[n] > top)
            {
                times = 0;
                top = D[n];
            }
            if (D[n] == top)
            {
                times++;
            }
        }
        return top * times;
    
    }

    static int Space_heavy(vector<int>& nums) {
        unordered_map<int, int> D;
        for (int n: nums)
            D[n]++;
        vector<PII> vp;
        for (auto & it: D)
            vp.push_back(it);
        sort(vp.begin(), vp.end(), cmp);
        int top = vp[0].second;
        int i = -1, res = 0;
        while (++i < (int)vp.size())
        {
            if (vp[i].second == top)
                res += top;
            else
                break;
        }
        return res;
    }

    static bool cmp(PII & l, PII & r)
    {
        return l.second > r.second;
    }
};
